src/vm/threads.cpp

```diff
@@ -1036,11 +1039,13 @@ HRESULT Thread::DetachThread(BOOL fDLLThreadDetach)
 #ifdef FEATURE_PERFTRACING
     // Before the thread dies, mark its buffers as no longer owned
     // so that they can be cleaned up after the thread dies.
-    EventPipeBufferList *pBufferList = m_pEventPipeBufferList.Load();
+    
+    EventPipeBufferList *pBufferList = GetThreadEventBufferList();
     if(pBufferList != NULL)
     {
         pBufferList->SetOwnedByThread(false);
```

Comment:
If the buffers aren't owned by the Thread object any longer then we should move the cleanup code to a path that applies to both managed and unmanaged threads. As it is, creating an unmanaged thread, logging from it, then destroying it would leak a buffer. CExecutionEngine::ThreadDetaching looks like a reasonable place though admittedly I am not that familiar with that portion of our code. @kouvel might be more knowledgable about it.

Reply:
It looks like CExecutionEngine::ThreadDetaching only applies to Thread objects and is called when the Thread object is being cleaned up. One way may be to add a destructor to ThreadLocalInfo, seems to work with __declspec(thread) but would need to ensure that it also works with clang. Another option is to keep the thread local info in Thread and create Thread objects. The Thread object is not necessarily tied to managed threads, that would be the m_exposedObject in Thread. For instance, wait threads in the thread pool (see ThreadpoolMgr::CreateWaitThread()) set up the thread to create a Thread object even though they would not run any managed code. I don't know the code well enough yet to tell when a Thread object is necessary vs not.

Reply:
Thanks Kount! I agree if we could do a destructor that seems very elegant.

@jkotas believed creating a Thread for those threads would be the harder of the options (https://github.com/dotnet/coreclr/issues/21380#issuecomment-447876984) but I don't have the GC knowledge to have a good opinion myself.